File: SecondYearCS/OOP/CivitasRB/lib/titulo_propiedad.rb
module CivitasRB
  require_relative 'jugador.rb'
  class TituloPropiedad
    #Setters/getters para las variables privadas necesarias.
    attr_accessor :numCasas, :numHoteles, :precioCompra,
                  :precioEdificar, :nombre, :Propietario, :alquilerBase
                
    #Constructor con parámetros para inicializar las variables.
    def initialize(nom, ab, fr,hb,pc, pe,intereses_hipoteca=1.1)
      @nombre = nom
      @alquilerBase = ab
      @factorRevalorizacion = fr
      @hipotecaBase = hb
      @precioCompra = pc
      @precioEdificar = pe
      @hipotecado = false
      @Propietario = nil
      @numCasas = 0
      @numHoteles = 0
      @@factorInteresesHipoteca = intereses_hipoteca
    end
    
    #Convertimos el estado de nuestra clase en una string.
    def to_s
      "El nombre de la propiedad es " + @nombre + ". El precio base " +
        "de alquiler y hipoteca es " + @alquilerBase.to_s + "€ " + @hipotecaBase.to_s
      + "€. El factor de valorizacion es " + @factorRevaloriacion.to_s +
        ". Por ultimo el precio de compra y edificacion es " + @precioCompra.to_s
      + "€ y " + @precioEdificar.to_s + "€."
    end
    
    #Verificamos si el param Jugador es realmente el propietario.
    def esEsteElPropietario(jugador)
      return @Propietario == jugador
    end
    
    #Utlizamos la formula del juego para calcular el precio del alquiler.
    def getPrecioAlquiler
      result = 0
      if(self.propietarioEncarcelado  && !@hipotecado)
        result = @alquilerBase*(1+(@numCasas*0.5)+(@numHoteles*2.5))
      end
      return result
    end
    
    #Utilizamos la formula del juego para calcular el precio de cancelar
    #El importe de la hipoteca
    # See also getImporteHipoteca.
    def getImporteCancelarHipoteca
       return (self.getImporteHipoteca * @@factorInteresesHipoteca)
    end
    
    #Utilizamos la formula del juego para calcular el importe de hipoteca.
    def getImporteHipoteca
      @hipotecaBase*(1+(@numCasas*0.5)+(@numHoteles*2.5))
    end
    
    #Si el jugador no es el propietario de la casilla
    #Entonces debemso tramitar el alquiler.
    #see self.getPrecioAlquiler y jugador.pagaAlquiler.
    def tramitarAlquiler(jugador)
      if(self.tienePropietario)
        if !self.esEsteElPropietario(jugador)
          jugador.pagaAlquiler(self.getPrecioAlquiler)
          @Propietario.recibe(self.getPrecioAlquiler)
        end
      end
    end
    
    #Verificamos si el propietario está o no encarcelado.
    def propietarioEncarcelado
      result = false
      if @Propietario != nil
        if @Propietario.encarcelado
          result = true
        end
      end
      return result
    end
    
    def cantidadCasasHoteles
      return @numCasas + @numHoteles
    end
    
    def getPrecioVenta
     return @precioCompra*(@numCasas+5*@numHoteles)*@precioEdificar*@factorRevalorizacion
    end
    
    #Debemos derruir un numero dado de casas si tenemos más casas que 
    #el parametro dado y además somos el propietario de la casilla.
    def derruirCasas(n, jugador)
      result = false
      if self.esEsteElPropietario(jugador) && @numCasas >= n
        @numCasas -= n
        result = true
      end
      return result 
    end
    
    #Si somos el propietario y además no estamos hipotecado
    #Entonces podemos vender nuestra propiedad 
    #See also jugador.recibe()
    def vender(jugador)
      result = false
      if self.esEsteElPropietario(jugador) && !@hipotecado
        jugador.recibe(self.getPrecioVenta)
        @Propietario = nil
        @numCasas = 0
        @numHoteles = 0
        result = true
      end
      return result
    end
    
    #Si no tiene propietario, y además el jugador pasado como parámetro
    #tiene más dinero que el necesario para comprarlo ( y puedeComprar).
    #entonces este pasa a ser el propietario de la casilla.
    def comprar(jugador)
      result = false
      if !self.tienePropietario
        @Propietario = jugador
        jugador.paga(@precioCompra)
        result = true
      end
      return result
    end
    
    #Si no hemos alcanzado el máximo de casas que el jugador
    #Puede construir, entonces construimos una casa más.
    def construirCasa(jugador)
      result = false
      if self.esEsteElPropietario(jugador)
        jugador.paga(@precioEdificar)
        @numCasas += 1
        result = true
      end
      return result
    end
    
    #Misma función que construir_casa pero con hoteles.
    def construirHotel(jugador)
      result = false
      if self.esEsteElPropietario(jugador)
        jugador.paga(@precioEdificar)
        @numHoteles += 1
        result = true
      end
      return result 
    end
    
    #Si somos el propietario, podemos hipotecar neustra propiedad.
    #See also jugador.recibe y self.getImporteHipoteca.
    def hipotecar(jugador)
      result = false
      if !@hipotecado && self.esEsteElPropietario(jugador)
        @hipotecado = true
        jugador.recibe(self.getImporteHipoteca)
        result = true
      end
      return result
    end
    
    #Si somos el propietario, podemos cancelar la hipoteca
    #see also jugador.paga y self.getImporteCancelarHipoteca.
    def cancelarHipoteca(jugador)
      result = false
      if self.esEsteElPropietario(jugador) && @hipotecado
        @hipotecado = false
        jugador.paga(self.getImporteCancelarHipoteca)
        result = true
      end
      return result
    end
    
    #Verificamos si la casilla ya posee propietario o no.
    def tienePropietario
      result = false
      if @Propietario != nil
        result = true
      end
      return result
    end
    
    def actualizaPropietarioPorConversion(jugador)
      @Propietario = jugador;
    end
    
  end
end

File: SecondYearCS/OOP/Civitas/src/civitas/TituloPropiedad.java
package civitas;

public class TituloPropiedad {
    
    private String nombre;
    private float alquilerBase, hipotecaBase, factorRevalorizacion,
               precioCompra, precioEdificar;
    private static final float factorInteresesHipoteca = 1.1f;
    private Jugador Propietario;
    private int numCasas;
    private int numHoteles;
    private boolean hipotecado;
            
    //Constructor con parámetros para inicializar todas las variables.
    TituloPropiedad(String nom,float ab, float fr, float hb,
                    float pc, float pe){
        this.nombre = nom;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        hipotecado = false;
        Propietario = null;
        numCasas = 0;
        numHoteles = 0;
        
    }
    
    /**
     * @brief SobreCarga del operador toString para devolver el estado la clase
     * @return Tipo String
     */
    @Override
    public String toString(){
       String result = "\n\tEl nombre de la propiedad es " + this.nombre
            + "\n\tHipotecaBase: "+ this.hipotecaBase + " €.\n\tEl factor de valorizacion es " + 
            this.factorRevalorizacion + "\n 1- precio de compra: " + precioCompra
               + "\n 2- precio de edificación: " + precioEdificar
               + "\n 3- alquiler base: " + alquilerBase;
       
       return result;
    }


    /**
    * @brief Si el propietario no está encarcelado y no hipotecado, calculamos
    * el precio del alquiler con la fórmula del juego
    * @see Jugador.isEncarcelado()
    * @return valor a pagar.
    */
    public float getPrecioAlquiler(){
        float result = 0;
        if ( !this.propietarioEncarcelado() && !hipotecado )
            result = (float) (this.alquilerBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
        return result;
    }
    
    /**
     * @brief Calculamos el importe de cancelar la hipoteca con la fórmula del 
     * juego y auxiliandonos del otro método de la clase. Devuelve el importe 
     * que se obtiene al hipotecar el título multiplicado por factorInteresesHipoteca
     * @see getImporteHipoteca()
     * @return precio a pagar.
     */
    float getImporteCancelarHipoteca(){
        return (float) ((float)this.getImporteHipoteca()*factorInteresesHipoteca);
    }

    /**
     * @brief Si @b this.Propietario no es igual al jugador ( parámetro ) entonces
     * debemos tramitar el alquiler a pagar
     * @see Jugador.pagaAlquiler(), getPrecioAlquiler()
     * @param jugador 
     */
    void tramitarAlquiler(Jugador jugador){
        if(this.tienePropietario()){
            if (!esEsteElPropietario(jugador)){
                jugador.pagaAlquiler(this.getPrecioAlquiler());
                this.Propietario.recibe(this.getPrecioAlquiler());
            }
        }
    }

    /**
     * @brief verificamos si el propietario esta encarcelado
     * @see Jugador.isEncarcelado
     * @return 
     */
    private boolean propietarioEncarcelado(){
        boolean result = false;
        if(this.Propietario != null){
            if(this.Propietario.isEncarcelado() )
                result = true;
        }
        return result;
    }

    /**
     * @brief Calculamos la suma del nº de casas y hoteles de la propiedad
     * @return El nº total de casas y hoteles
     */
    int cantidadCasasHoteles(){
        return numCasas+numHoteles;
    }  

    /**
     * @brief Aplicamos la fórmula del juego para Obtener el precio de venta.
     * @return precio a pagar
     */
    private float getPrecioVenta(){
        return (precioCompra+(numCasas+5*numHoteles)
                *precioEdificar*factorRevalorizacion);
    }

    /**
     * @brief Si podemos, destruimos n casas que poseemos
     * @param n número de casas a derruir
     * @param jugador propietario de las casas
     * @return succeso / error
     */
    boolean derruirCasas(int n, Jugador jugador){
        boolean result = false;
        if (esEsteElPropietario(jugador) && numCasas >= n){
            numCasas -= n;
            result = true;
        }
        return result;
    }

    /**
     * @brief vendemos nuestra propiedad
     * @param jugador el propietario
     * @return succeso / error
     */
    boolean vender(Jugador jugador){
        boolean result = false;
        if(esEsteElPropietario(jugador) && !hipotecado){
            jugador.recibe(this.getPrecioVenta());
            this.Propietario = null;
            this.numCasas = 0;
            this.numHoteles = 0;
            result = true;
        }
        return result;
    }

    /**
     * @brief Si somos el propietario y estamos hipotecado entonces este recibe 
     * el imoprte de cancelar la hipoteca.
     * @see getImporteCancelarHipoteca(), Jugador.paga()+
     * @param jugador
     * @return 
     */
    boolean cancelarHipoteca(Jugador jugador){
         boolean result = false;
        if(esEsteElPropietario(jugador) && this.hipotecado){
            this.hipotecado = false;
            jugador.paga(this.getImporteCancelarHipoteca());
            result = true;
        }
        return result;
    }

    /**
     * @brief Si el jugador es el propietario entonces recibe el importe de Hipoteca
     * @see getImporteHipoteca(), Jugador.recibe()
     * @param jugador propietario
     * @return succeso / errors
     */
    boolean hipotecar(Jugador jugador){
        boolean result = false;
        if(!hipotecado && esEsteElPropietario(jugador)){
            this.hipotecado = true;
            jugador.recibe(this.getImporteHipoteca());
            result = true;
        }
        return result;
    }

    /**
     * @brief Si tenemos dinero suficiente entonces podemos comprar la propiedad.
     * @param jugador comprador
     * @return succeso / error 
     */
     boolean comprar(Jugador jugador){
        boolean result = false;
        if(!this.tienePropietario()){
            this.Propietario = jugador;
            jugador.paga(this.precioCompra);
            result = true;
        }
        return result;
    }

    /**
     * @brief Si somos el propietario entonces podemos construir más casas
     * @param jugador propietario
     * @return succeso / error
     */
    boolean construirCasa(Jugador jugador){
        boolean result = false;
        if(esEsteElPropietario(jugador)){
            jugador.paga(precioEdificar);
            numCasas++;
            result = true;
        }
        return result;
    }

    /**
     * @brief Similar al método para construir casas
     * @see construirCasa()
     * @param jugador
     * @return 
     */
    boolean construirHotel(Jugador jugador){
        boolean result = false;
        if(esEsteElPropietario(jugador)){
            this.derruirCasas(numCasas, jugador);
            jugador.paga(precioEdificar);
            numHoteles++;
            result = true;
        }
        return result;
    }

    /**
     * @brief Verifica si el parámetro es el @b this.Propietario de la clase.
     * @param jugador Jugador a comparar con
     * @return verdadero / falso
     */
    private boolean esEsteElPropietario(Jugador jugador){
        return this.Propietario == jugador;
    }
    
    /**
     * @brief Convertimos el @b this.Propietario en el nuevo valor pasado
     * @param jugador nuevo valor de @b this.Propietario.
     */
    void actualizaPropietarioPorConversion(Jugador jugador){
        this.Propietario = jugador;
    }
    
    
    /**
     * @brief Aplicamos la fórmula del juego para calcular el importe de Hipoteca
     * @return precio a pagar 
     */
    private float getImporteHipoteca(){
        return (float) (this.hipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
    }
    

    public String getNombre() {
        return nombre;
    }

    float getPrecioCompra() {
        return precioCompra;
    }

    float getPrecioEdificar() {
        return precioEdificar;
    }

    boolean tienePropietario(){
        boolean result = false;
        if(Propietario!=null){
            result = true;
        }
        return result;
    }
        
    Jugador getPropietario() {
        return Propietario;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getNumHoteles() {
        return numHoteles;
    }
    
    public boolean getHipotecado(){
        return hipotecado;
    }
    
    public float getAlquilerBase(){
        return alquilerBase;
    }
    
}
